wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/activity/WorkspaceActivityManager.java

```
@@ -0,0 +1,148 @@
+                        break;
+                    case STOPPED:
+                        activeWorkspaces.remove(event.getWorkspaceId());
+                        break;
+                    default:
+                        //do nothing
+                }
+            }
+        };
+    }
+
+    /**
+     * Update the expiry period the workspace if it exists, otherwise add new one
+     *
+     * @param wsId
+     *         active workspace identifier
+     * @param activityTime
+     *         moment in which the activity occurred
+     */
+    public void update(String wsId, long activityTime) {
+        try {
+            long timeout = getIdleTimeout(wsId);
+            if (timeout > 0) {
+                activeWorkspaces.put(wsId, activityTime + timeout);
+            }
+        } catch (NotFoundException | ServerException e) {
+            LOG.error(e.getLocalizedMessage(), e);
+        }
+    }
+
+    protected long getIdleTimeout(String workspaceId) throws NotFoundException, ServerException {
+        if (timeout > 0) {
+            return timeout;
+        } else {
+            return -1;
+        }
+    }
+
+    @ScheduleRate(periodParameterName = "stop.workspace.scheduler.period")
+    private void invalidate() {
+        final long currentTime = System.currentTimeMillis();
+        for (Map.Entry<String, Long> workspaceExpireEntry : activeWorkspaces.entrySet()) {
+            if (workspaceExpireEntry.getValue() <= currentTime) {
+                try {
+                    String workspaceId = workspaceExpireEntry.getKey();
+                    Workspace workspace = workspaceManager.getWorkspace(workspaceId);
+                    workspace.getAttributes().put(WORKSPACE_STOPPED_BY, ACTIVITY_CHECKER);
+                    workspaceManager.updateWorkspace(workspaceId, workspace);
+                    workspaceManager.stopWorkspace(workspaceId);
+                } catch (NotFoundException e) {
+                    LOG.info("Workspace already stopped");
```

Please ignore this exception - there is no need to log it